Mathlib/Algebra/Lie/Basic.lean

```diff
@@ -0,0 +1,1144 @@
+#align zero_lie zero_lie
+
+@[simp]
+theorem lie_self : ⁅x, x⁆ = 0 :=
+  LieRing.lie_self x
+#align lie_self lie_self
+
+instance lieRingSelfModule : LieRingModule L L :=
+  { (inferInstance : LieRing L) with }
+#align lie_ring_self_module lieRingSelfModule
+
+@[simp]
+theorem lie_skew : -⁅y, x⁆ = ⁅x, y⁆ := by
+  have h : ⁅x + y, x⁆ + ⁅x + y, y⁆ = 0 := by rw [← lie_add]; apply lie_self
+  simpa [neg_eq_iff_add_eq_zero] using h
+#align lie_skew lie_skew
+
+/-- Every Lie algebra is a module over itself. -/
+instance lieAlgebraSelfModule : LieModule R L L
+    where
+  smul_lie t x m := by rw [← lie_skew, ← lie_skew x m, LieAlgebra.lie_smul, smul_neg]
+  lie_smul := by apply LieAlgebra.lie_smul
+#align lie_algebra_self_module lieAlgebraSelfModule
+
+@[simp]
+theorem neg_lie : ⁅-x, m⁆ = -⁅x, m⁆ := by
+  rw [← sub_eq_zero, sub_neg_eq_add, ← add_lie]
+  simp
+#align neg_lie neg_lie
+
+@[simp]
+theorem lie_neg : ⁅x, -m⁆ = -⁅x, m⁆ := by
+  rw [← sub_eq_zero, sub_neg_eq_add, ← lie_add]
+  simp
+#align lie_neg lie_neg
+
+@[simp]
+theorem sub_lie : ⁅x - y, m⁆ = ⁅x, m⁆ - ⁅y, m⁆ := by simp [sub_eq_add_neg]
+#align sub_lie sub_lie
+
+@[simp]
+theorem lie_sub : ⁅x, m - n⁆ = ⁅x, m⁆ - ⁅x, n⁆ := by simp [sub_eq_add_neg]
+#align lie_sub lie_sub
+
+@[simp]
+theorem nsmul_lie (n : ℕ) : ⁅n • x, m⁆ = n • ⁅x, m⁆ :=
+  AddMonoidHom.map_nsmul ⟨⟨fun x : L => ⁅x, m⁆, zero_lie m⟩, fun _ _ => add_lie _ _ _⟩ _ _
+#align nsmul_lie nsmul_lie
+
+@[simp]
+theorem lie_nsmul (n : ℕ) : ⁅x, n • m⁆ = n • ⁅x, m⁆ :=
+  AddMonoidHom.map_nsmul ⟨⟨fun m : M => ⁅x, m⁆, lie_zero x⟩, fun _ _ => lie_add _ _ _⟩ _ _
+#align lie_nsmul lie_nsmul
+
+@[simp]
+theorem zsmul_lie (a : ℤ) : ⁅a • x, m⁆ = a • ⁅x, m⁆ :=
+  AddMonoidHom.map_zsmul ⟨⟨fun x : L => ⁅x, m⁆, zero_lie m⟩, fun _ _ => add_lie _ _ _⟩ _ _
+#align zsmul_lie zsmul_lie
+
+@[simp]
+theorem lie_zsmul (a : ℤ) : ⁅x, a • m⁆ = a • ⁅x, m⁆ :=
+  AddMonoidHom.map_zsmul ⟨⟨fun m : M => ⁅x, m⁆, lie_zero x⟩, fun _ _ => lie_add _ _ _⟩ _ _
+#align lie_zsmul lie_zsmul
+
+@[simp]
+theorem lie_lie : ⁅⁅x, y⁆, m⁆ = ⁅x, ⁅y, m⁆⁆ - ⁅y, ⁅x, m⁆⁆ := by rw [leibniz_lie, add_sub_cancel]
+#align lie_lie lie_lie
+
+theorem lie_jacobi : ⁅x, ⁅y, z⁆⁆ + ⁅y, ⁅z, x⁆⁆ + ⁅z, ⁅x, y⁆⁆ = 0 := by
+  rw [← neg_neg ⁅x, y⁆, lie_neg z, lie_skew y x, ← lie_skew, lie_lie]
+  abel
+#align lie_jacobi lie_jacobi
+
+instance LieRing.intLieAlgebra : LieAlgebra ℤ L where lie_smul n x y := lie_zsmul x y n
+#align lie_ring.int_lie_algebra LieRing.intLieAlgebra
+
+-- Porting note: TODO Erase this line. Needed because we don't have η for classes. (lean4#2074)
+attribute [-instance] Ring.toNonAssocRing
+
+instance : LieRingModule L (M →ₗ[R] N) where
+  bracket x f :=
+    { toFun := fun m => ⁅x, f m⁆ - f ⁅x, m⁆
+      map_add' := fun m n => by
+        simp only [lie_add, LinearMap.map_add]
+        abel
+      map_smul' := fun t m => by
+        simp only [smul_sub, LinearMap.map_smul, lie_smul, RingHom.id_apply] }
+  -- Porting note: TODO fill in these sorrys
```

Comment:
These `sorry`s can be closed by adding the following lemma:
```lean
@[simp] lemma LinearMap.coe_mk' {R M N : Type _} [Semiring R] [AddCommMonoid M] [AddCommMonoid N]
  [Module R M] [Module R N] {f : AddHom M N} (h) (m : M) :
  (LinearMap.mk f h : M →ₗ[R] N) m = f m :=
rfl
```
and then adding `LinearMap.mk_apply, AddHom.coe_mk` to the list of lemmas supplied to `simp only` in each of the missing proofs below (and also changing `abel_nf` to `abel`).

I suppose the point is that in Lean 4 we only have what we call "new style" structures in Lean 3 and so need more lemmas of this shape but I have some homework to do before I propose this as the solution (and anyway it would have to be generalised to the semilinear setting).

Reply:
On second thought, perhaps the point is that [https://github.com/leanprover-community/mathlib4/blob/065e9956724008d1059577700fd676eaf25f35d5/Mathlib/Algebra/Module/LinearMap.lean#L272](LinearMap.coe_mk) should be changed from `f : M →+ M₃` to `f : AddHom M M₃`.